### PR TITLE
Change to use compact rcf model config

### PR DESCRIPTION
**Pull request by kaituo**
Note: since there are inter-file references, I split a big PR into multiple smaller PRs based on files to save time (1.1 cut off is due at the end of this week). The build will fail due to missing dependencies. I will use that PR just for review. will not merge it. Will have a big PR in the end and merge once after all review PRs get approved. For the complete code, check https://github.com/kaituo/anomaly-detection-1/tree/compactRCF2

### Description
The PR made the following changes.

1)Change to use compact rcf, and float 32 precision.
2)Use zero bounding box cache for real-time detectors. The change means every RCF update/score has to recreate all of the bounding boxes. However, the change is fine for real-time detectors as we won't access the same forest a second time, at least for another minute, as the lowest resolution of a detector interval is one minute.
3)Use bounding box cache ratio 1 (i.e., all bounding box cache access is a hit) for historical detectors. Having a full bounding box intact is the original behavior of RCF 1.0. In batch mode, recreating all of the bounding boxes all the time impacts performance.  Thus, we keep the original behavior and make the bounding box cache ratio to be 1.
4)Enable shingleSize optimization. This optimization will try to save space opportunistically. Usually, we use it as shingleSize(dimension). When a new point comes in, we will look at the point store if there is any overlapping. Say the previously-stored vector is x1, x2, x3, x4, now we add x3, x4, x5, x6. RCF will recognize overlapping x3, x4, and only store x5, x6.

Testing done:
1. tested the change does not impact existing functionality.
 
### Issues Resolved
https://github.com/opensearch-project/anomaly-detection/issues/97
 
### Check List
- [ x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java [hunk 1]
```diff
@@ -40,18 +40,22 @@
...
 import org.opensearch.ad.settings.AnomalyDetectorSettings;
 
 import com.amazon.randomcutforest.RandomCutForest;
+import com.amazon.randomcutforest.config.Precision;
 
 /**
  * AD batch task cache which will hold RCF, threshold model, shingle and training data.
  */
 public class ADBatchTaskCache {
+    private static final Logger LOG = LogManager.getLogger(ADBatchTaskCache.class);
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:58]: Seems this `LOG` not being used

File: src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java [hunk 2]
```diff
@@ -70,14 +74,20 @@ protected ADBatchTaskCache(ADTask adTask) {
...
-            .lambda(TIME_DECAY)
+            .timeDecay(TIME_DECAY)
             .sampleSize(NUM_SAMPLES_PER_TREE)
             .outputAfter(NUM_MIN_SAMPLES)
             .parallelExecutionEnabled(false)
+            .compact(true)
+            .precision(Precision.FLOAT_32)
+            .boundingBoxCacheFraction(AnomalyDetectorSettings.BATCH_BOUNDING_BOX_CACHE_RATIO)
+            // same with dimension for opportunistic memory saving
+            .shingleSize(dimensions)
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: Why not use `detector.getShingleSize()` as RCF's shingle size? If enabled feature size is 5, and detector single size is 8, then the RCF's shingle size will be 40?

Comment (ylwu-amzn) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: Seems `shingleSize` is new parameter of RCF, what's the difference with `dimension`? I don't see `EntityColdStarter.java` set `shingleSize` when create RCF.

File: src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java (hunk 1 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:58]: will remove

File: src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: rcf's shingle size is different from our detector's shingle size.  " If enabled feature size is 5, and detector single size is 8, then the RCF's shingle size will be 40" that's correct.

Comment (kaituo) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: We do use shingleSize at line 344: https://github.com/opensearch-project/anomaly-detection/pull/143/files#diff-235791084d307db4d86529d4f484ef3e6f213af48907cc003298b8e0e2211192R344

Comment (kaituo) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: I explained the shingleSize parameter in the overview: "Enable shingleSize optimization. This optimization will try to save space opportunistically. Usually, we use it as shingleSize(dimension). When a new point comes in, we will look at the point store if there is any overlapping. Say the previously-stored vector is x1, x2, x3, x4, now we add x3, x4, x5, x6. RCF will recognize overlapping x3, x4, and only store x5, x6."

Comment (kaituo) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: will add this as a comment as well

Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: Wait! For RCF 2.0 we would need to set dimensions = 40 and shinglesize =8. We can set other values of shingle size -- the code will match opportunistically but we may not see any savings. For shingle size 8, we will get a match and only add 5 new values per item

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java [hunk 3]
```diff
@@ -331,9 +334,14 @@ private void trainModelFromDataSegments(List<double[][]> dataPoints, Entity enti
...
             .numberOfTrees(numberOfTrees)
-            .lambda(rcfTimeDecay)
+            .timeDecay(rcfTimeDecay)
             .outputAfter(numMinSamples)
             .parallelExecutionEnabled(false)
+            .compact(true)
+            .precision(Precision.FLOAT_32)
+            .boundingBoxCacheFraction(AnomalyDetectorSettings.REAL_TIME_BOUNDING_BOX_CACHE_RATIO)
+            // same with dimension for opportunistic memory saving
+            .shingleSize(rcfNumFeatures)
```
Comment (sudiptoguha) [inline src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:344]: I think it should be ShingleSize?

File: src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java (hunk 2 above)
Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: so just ".shingleSize(detector.getShingleSize())"

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java (hunk 3 above)
Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:344]: that is ".shingleSize(shingleSize)"

File: src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java (hunk 2 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: right, will change

@ylwu-amzn  shingleSize in rcf and AD are the same.  Sorry for the confusion.  My explanation about the shingleSize parameter still holds

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java (hunk 3 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:344]: changed.

File: src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java (hunk 2 above)
Comment (ylwu-amzn) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: Cool. That part looks clear now.

For this 
> Enable shingleSize optimization. This optimization will try to save space opportunistically. Usually, we use it as shingleSize(dimension). When a new point comes in, we will look at the point store if there is any overlapping. Say the previously-stored vector is x1, x2, x3, x4, now we add x3, x4, x5, x6. RCF will recognize overlapping x3, x4, and only store x5, x6.

Can you add it to specific line of code? Still not quite understand which part in this PR does this optimization.

Comment (kaituo) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java:90]: I did.  The changed code looks like

```
rcfModel = RandomCutForest
            .builder()
            .dimensions(dimensions)
            .numberOfTrees(numberOfTrees)
            .timeDecay(TIME_DECAY)
            .sampleSize(NUM_SAMPLES_PER_TREE)
            .outputAfter(NUM_MIN_SAMPLES)
            .parallelExecutionEnabled(false)
            .compact(true)
            .precision(Precision.FLOAT_32)
            .boundingBoxCacheFraction(AnomalyDetectorSettings.BATCH_BOUNDING_BOX_CACHE_RATIO)
            // same with dimension for opportunistic memory saving
            // Usually, we use it as shingleSize(dimension). When a new point comes in, we will
            // look at the point store if there is any overlapping. Say the previously-stored
            // vector is x1, x2, x3, x4, now we add x3, x4, x5, x6. RCF will recognize
            // overlapping x3, x4, and only store x5, x6.
            .shingleSize(shingleSize)
            .build();
```

Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/task/ADBatchTaskCache.java]: Looks good.

**Review by ylwu-amzn — APPROVED** [review]

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java (hunk 3 above)
Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/ml/EntityColdStarter.java]: Not seeing change though ... should I refresh/do something else?

**Review by sudiptoguha — APPROVED** [review]

**Pull request closed without merging**